### PR TITLE
Run3-hgx219 Convert algorithm for DDHGCalTBModuleX

### PR DESCRIPTION
**Pull request by bsunanda**
#### PR description:

Convert the DD algorithm DDHGCalTBModuleX to DD4Hep format

#### PR validation:

Test to be completed using the script in Geometry/HGCalCommonData/test/python

#### if this PR is a backport please specify the original PR:

Nothing special

### CODE REVIEW

**Comment by kpedro88** [comment]
please test

**Comment by cvuosalo** [comment]
+1

File: Geometry/HGCalCommonData/plugins/dd4hep/DDHGCalTBModuleX.cc [hunk 1]
```diff
@@ -0,0 +1,378 @@
...
+                       const std::vector<std::string>& names,
+                       const std::vector<std::string>& materials,
+                       std::vector<int>& copyNumber,
+                       const std::vector<double>& layerThick,
+                       const double& absorbW,
+                       const double& absorbH,
+                       const double& waferTot,
+                       const double& rMax,
+                       const double& rMaxFine,
+#ifdef EDM_ML_DEBUG
```
[the remark this answers is not in the document]
Comment (bsunanda) [reply Geometry/HGCalCommonData/plugins/dd4hep/DDHGCalTBModuleX.cc]: Please let me know if you want the 2 variables to be kept all the time in the calling sequence or not.

**Comment by bsunanda** [comment]
Hi Kevin

Could you please explain what is expected from me - should I always include these two variables in the calling sequence?

Sunanda
________________________________

File: Geometry/HGCalCommonData/plugins/dd4hep/DDHGCalTBModuleX.cc (hunk 1 above)
[the remark this answers is not in the document]
Comment (kpedro88) [reply Geometry/HGCalCommonData/plugins/dd4hep/DDHGCalTBModuleX.cc]: Yes, it seems like the safer approach with little downside.

**Comment by ianna** [comment]
hold

@bsunanda - please, see my comments

File: Geometry/HGCalCommonData/plugins/dd4hep/DDHGCalTBModuleX.cc [hunk 2]
```diff
@@ -0,0 +1,378 @@
...
+  for (unsigned int i = 0; i < blockThick.size(); i++) {
+    double zo = zi + blockThick[i];
+    std::string name = idName + "Block" + std::to_string(i);
+#ifdef EDM_ML_DEBUG
+    edm::LogVerbatim("HGCalGeom") << "DDHGCalTBModuleX: Block " << i << ":" << name << " z " << convertCmToMm(zi) << ":"
+                                  << convertCmToMm(zo) << " R " << convertCmToMm(rMaxB) << " T "
+                                  << convertCmToMm(blockThick[i]);
+#endif
+    dd4hep::Material matter = ns.material(genMat);
+    dd4hep::Solid solid = dd4hep::Tube(0.5 * blockThick[i], 0, rMaxB, 0.0, 2._pi);
```
Comment (ianna) [inline Geometry/HGCalCommonData/plugins/dd4hep/DDHGCalTBModuleX.cc]:
```
//  Tube(rmin, rmax, dz, startPhi, startPhi + deltaPhi)
-    dd4hep::Solid solid = dd4hep::Tube(0.5 * blockThick[i], 0, rMaxB, 0.0, 2._pi);
+    dd4hep::Solid solid = dd4hep::Tube(0, rMaxB, 0.5 * blockThick[i], 0.0, 2._pi);
```
@bsunanda - The Tube API has changed and this one was causing the voxelization problem:
```
----- Begin Fatal Exception 16-Sep-2019 10:25:05 CEST-----------------------
An exception of category 'StdException' occurred while
   [0] Processing  Event run: 1 lumi: 1 event: 1 stream: 0
   [1] Running path 'p'
   [2] Calling method for module DDTestDumpFile/'testDump'
   [3] Using EventSetup component DDDetectorESProducer/'' to make data DDDetector/'DDHGCalTBModuleX' in record GeometryFileRcd
Exception Message:
A std::exception was thrown.
An exception of category 'FatalRootError' occurred.
   Additional Info:
      [a] Fatal Root Error: @SUB=TGeoVoxelFinder::SortAll
Cannot voxelize HGCalEE :less than 2 boundaries on X


dd4hep: while parsing /build/yana/DD4HEP/Sunanda/CMSSW_11_0_X_2019-09-12-2300/src/Geometry/HGCalCommonData/data/dd4hep/cms-test-ddhgcalTBModuleX-algorithm.xml
dd4hep: with plugin:DD4hep_CompactLoader
----- End Fatal Exception -------------------------------------------------
```

File: Geometry/HGCalCommonData/plugins/dd4hep/DDHGCalTBModuleX.cc [hunk 3]
```diff
@@ -0,0 +1,378 @@
...
+    ns.addSolidNS(ns.prepend(name), solid);
+    dd4hep::Volume glog = dd4hep::Volume(solid.name(), solid, matter);
+    double zz = zi + 0.5 * blockThick[i];
+    dd4hep::Position r1(0, 0, zz);
+    dd4hep::Rotation3D rot;
+    parent.placeVolume(glog, i, dd4hep::Transform3D(rot, r1));
```
Comment (ianna) [inline Geometry/HGCalCommonData/plugins/dd4hep/DDHGCalTBModuleX.cc]:
```
module.placeVolume(glog, copy, r1);
```

File: Geometry/HGCalCommonData/plugins/dd4hep/DDHGCalTBModuleX.cc [hunk 4]
```diff
@@ -0,0 +1,378 @@
...
+                                  << convertCmToMm(zo) << " R " << convertCmToMm(rMaxB) << " T "
+                                  << convertCmToMm(blockThick[i]);
+#endif
+    dd4hep::Material matter = ns.material(genMat);
+    dd4hep::Solid solid = dd4hep::Tube(0.5 * blockThick[i], 0, rMaxB, 0.0, 2._pi);
+    ns.addSolidNS(ns.prepend(name), solid);
+    dd4hep::Volume glog = dd4hep::Volume(solid.name(), solid, matter);
+    double zz = zi + 0.5 * blockThick[i];
+    dd4hep::Position r1(0, 0, zz);
+    dd4hep::Rotation3D rot;
```
Comment (ianna) [inline Geometry/HGCalCommonData/plugins/dd4hep/DDHGCalTBModuleX.cc]: please remove

File: Geometry/HGCalCommonData/plugins/dd4hep/DDHGCalTBModuleX.cc [hunk 5]
```diff
@@ -0,0 +1,378 @@
...
+        ns.addSolidNS(ns.prepend(name), solid);
+        glog = dd4hep::Volume(solid.name(), solid, matter);
+#ifdef EDM_ML_DEBUG
+        edm::LogVerbatim("HGCalGeom") << "DDHGCalTBModuleX: " << solid.name() << " box of dimension "
+                                      << convertCmToMm(absorbW) << ":" << convertCmToMm(absorbH) << ":"
+                                      << convertCmToMm(0.5 * layerThick[ii]);
+#endif
+        dd4hep::Position r1(0, 0, zz);
+        dd4hep::Rotation3D rot;
+        module.placeVolume(glog, copy, dd4hep::Transform3D(rot, r1));
```
Comment (ianna) [inline Geometry/HGCalCommonData/plugins/dd4hep/DDHGCalTBModuleX.cc]:
```
module.placeVolume(glog, copy, r1);
```

File: Geometry/HGCalCommonData/plugins/dd4hep/DDHGCalTBModuleX.cc [hunk 6]
```diff
@@ -0,0 +1,378 @@
...
+        dd4hep::Solid solid = dd4hep::Box(absorbW, absorbH, 0.5 * layerThick[ii]);
+        ns.addSolidNS(ns.prepend(name), solid);
+        glog = dd4hep::Volume(solid.name(), solid, matter);
+#ifdef EDM_ML_DEBUG
+        edm::LogVerbatim("HGCalGeom") << "DDHGCalTBModuleX: " << solid.name() << " box of dimension "
+                                      << convertCmToMm(absorbW) << ":" << convertCmToMm(absorbH) << ":"
+                                      << convertCmToMm(0.5 * layerThick[ii]);
+#endif
+        dd4hep::Position r1(0, 0, zz);
+        dd4hep::Rotation3D rot;
```
Comment (ianna) [inline Geometry/HGCalCommonData/plugins/dd4hep/DDHGCalTBModuleX.cc]: please remove

File: Geometry/HGCalCommonData/plugins/dd4hep/DDHGCalTBModuleX.cc [hunk 7]
```diff
@@ -0,0 +1,378 @@
...
+              bool cornerAll(true);
+              for (int k = 0; k < 6; ++k) {
+                double rpos = std::sqrt(xc[k] * xc[k] + yc[k] * yc[k]);
+                if (rpos > rMax)
+                  cornerAll = false;
+              }
+              if (cornerAll) {
+                double rpos = std::sqrt(xpos * xpos + ypos * ypos);
+                dd4hep::Position tran(xpos, ypos, zz);
+                dd4hep::Rotation3D rotation;
```
Comment (ianna) [inline Geometry/HGCalCommonData/plugins/dd4hep/DDHGCalTBModuleX.cc]: please remove

File: Geometry/HGCalCommonData/plugins/dd4hep/DDHGCalTBModuleX.cc [hunk 8]
```diff
@@ -0,0 +1,378 @@
...
+                  copyx += 10000;
+                if (nr < 0)
+                  copyx += 100000;
+                if (layerSense[ly] == 1) {
+                  dd4hep::Solid solid = ns.solid(covers[0]);
+                  std::string name0 = name + "M" + std::to_string(copyx);
+                  if (strchr(name0.c_str(), NAMESPACE_SEP) == nullptr)
+                    name0 = ns.name() + name0;
+                  dd4hep::Volume glog1 = dd4hep::Volume(name0, solid, matter);
+                  module.placeVolume(glog1, copy, dd4hep::Transform3D(rotation, tran));
```
Comment (ianna) [inline Geometry/HGCalCommonData/plugins/dd4hep/DDHGCalTBModuleX.cc]:
```
module.placeVolume(glog1, copy, tran);
```

File: Geometry/HGCalCommonData/plugins/dd4hep/DDHGCalTBModuleX.cc [hunk 9]
```diff
@@ -0,0 +1,378 @@
...
+#ifdef EDM_ML_DEBUG
+                  edm::LogVerbatim("HGCalGeom")
+                      << "DDHGCalTBModuleX: " << glog1.name() << " number " << copy << " positioned in "
+                      << module.name() << " at " << tran << " with " << rotation;
+#endif
+                  dd4hep::Volume glog2 = (rpos < rMaxFine) ? ns.volume(wafers[0]) : ns.volume(wafers[1]);
+                  dd4hep::Position tran1;
+                  glog1.placeVolume(glog2, copyx, dd4hep::Transform3D(rotation, tran1));
```
Comment (ianna) [inline Geometry/HGCalCommonData/plugins/dd4hep/DDHGCalTBModuleX.cc]:
```
glog1.placeVolume(glog2, copyx);
```

File: Geometry/HGCalCommonData/plugins/dd4hep/DDHGCalTBModuleX.cc [hunk 10]
```diff
@@ -0,0 +1,378 @@
...
+                    name0 = ns.name() + name0;
+                  dd4hep::Volume glog1 = dd4hep::Volume(name0, solid, matter);
+                  module.placeVolume(glog1, copy, dd4hep::Transform3D(rotation, tran));
+#ifdef EDM_ML_DEBUG
+                  edm::LogVerbatim("HGCalGeom")
+                      << "DDHGCalTBModuleX: " << glog1.name() << " number " << copy << " positioned in "
+                      << module.name() << " at " << tran << " with " << rotation;
+#endif
+                  dd4hep::Volume glog2 = (rpos < rMaxFine) ? ns.volume(wafers[0]) : ns.volume(wafers[1]);
+                  dd4hep::Position tran1;
```
Comment (ianna) [inline Geometry/HGCalCommonData/plugins/dd4hep/DDHGCalTBModuleX.cc]: please remove

File: Geometry/HGCalCommonData/plugins/dd4hep/DDHGCalTBModuleX.cc [hunk 11]
```diff
@@ -0,0 +1,378 @@
...
+                  edm::LogVerbatim("HGCalGeom")
+                      << "DDHGCalTBModuleX: " << glog2.name() << " number " << copyx << " positioned in "
+                      << glog1.name() << " at " << tran1 << " with " << rotation;
+                  if (layerSense[ly] == 1)
+                    copies.insert(copy);
+#endif
+                } else {
+                  dd4hep::Volume glog2 = ns.volume(covers[layerSense[ly] - 1]);
+                  copyx += (copy * 1000000);
+                  module.placeVolume(glog2, copyx, dd4hep::Transform3D(rotation, tran));
```
Comment (ianna) [inline Geometry/HGCalCommonData/plugins/dd4hep/DDHGCalTBModuleX.cc]:
```
module.placeVolume(glog2, copyx, tran);
```

File: Geometry/HGCalCommonData/plugins/dd4hep/DDHGCalTBModuleX.cc [hunk 12]
```diff
@@ -0,0 +1,378 @@
+/*
+ * DDHGCalTBModuleX.cc
+ *
+ *  Created on: 27-Aug-2019
+ *      Author: S. Banerjee
+ */
+
+#include <unordered_set>
+
+#include "DataFormats/Math/interface/GeantUnits.h"
```
Comment (ianna) [inline Geometry/HGCalCommonData/plugins/dd4hep/DDHGCalTBModuleX.cc]:
```
#include "DataFormats/Math/interface/CMSUnits.h"
```

File: Geometry/HGCalCommonData/plugins/dd4hep/DDHGCalTBModuleX.cc [hunk 13]
```diff
@@ -0,0 +1,378 @@
...
+#include "DD4hep/DetFactoryHelper.h"
+#include "DetectorDescription/Core/interface/DDSplit.h"
+#include "DetectorDescription/DDCMS/interface/DDPlugins.h"
+#include "FWCore/MessageLogger/interface/MessageLogger.h"
+#include "FWCore/Utilities/interface/Exception.h"
+
+//#define EDM_ML_DEBUG
+using namespace geant_units::operators;
```
Comment (ianna) [inline Geometry/HGCalCommonData/plugins/dd4hep/DDHGCalTBModuleX.cc]:
```
-using namespace geant_units::operators;
+using namespace cms_units::operators;
```

**Review by ianna — CHANGES_REQUESTED** [review]
With the proposed changes the test runs fine.

**Comment by ianna** [comment]
unhold

**Comment by ianna** [comment]
-1

File: Geometry/HGCalCommonData/plugins/dd4hep/DDHGCalTBModuleX.cc (hunk 1 above)
[the remark this answers is not in the document]
Comment (bsunanda) [reply Geometry/HGCalCommonData/plugins/dd4hep/DDHGCalTBModuleX.cc]: Removed the ifdef in the call

File: Geometry/HGCalCommonData/plugins/dd4hep/DDHGCalTBModuleX.cc (hunk 2 above)
Comment (bsunanda) [reply Geometry/HGCalCommonData/plugins/dd4hep/DDHGCalTBModuleX.cc]: Thanks Yana - made the correction

File: Geometry/HGCalCommonData/plugins/dd4hep/DDHGCalTBModuleX.cc (hunk 3 above)
Comment (bsunanda) [reply Geometry/HGCalCommonData/plugins/dd4hep/DDHGCalTBModuleX.cc]: Thanks Yana - made the correction

File: Geometry/HGCalCommonData/plugins/dd4hep/DDHGCalTBModuleX.cc (hunk 4 above)
Comment (bsunanda) [reply Geometry/HGCalCommonData/plugins/dd4hep/DDHGCalTBModuleX.cc]: Thanks Yana - made the correction

File: Geometry/HGCalCommonData/plugins/dd4hep/DDHGCalTBModuleX.cc (hunk 5 above)
Comment (bsunanda) [reply Geometry/HGCalCommonData/plugins/dd4hep/DDHGCalTBModuleX.cc]: Thanks Yana - made the correction

File: Geometry/HGCalCommonData/plugins/dd4hep/DDHGCalTBModuleX.cc (hunk 6 above)
Comment (bsunanda) [reply Geometry/HGCalCommonData/plugins/dd4hep/DDHGCalTBModuleX.cc]: Thanks Yana - made the correction

File: Geometry/HGCalCommonData/plugins/dd4hep/DDHGCalTBModuleX.cc (hunk 9 above)
Comment (bsunanda) [reply Geometry/HGCalCommonData/plugins/dd4hep/DDHGCalTBModuleX.cc]: Thanks Yana - made the correction

File: Geometry/HGCalCommonData/plugins/dd4hep/DDHGCalTBModuleX.cc (hunk 11 above)
Comment (bsunanda) [reply Geometry/HGCalCommonData/plugins/dd4hep/DDHGCalTBModuleX.cc]: Thanks Yana - made the correction

File: Geometry/HGCalCommonData/plugins/dd4hep/DDHGCalTBModuleX.cc (hunk 12 above)
Comment (bsunanda) [reply Geometry/HGCalCommonData/plugins/dd4hep/DDHGCalTBModuleX.cc]: I need he GeantUnits.h for convertCMtoMM which is not in CMSUnits

File: Geometry/HGCalCommonData/plugins/dd4hep/DDHGCalTBModuleX.cc (hunk 13 above)
Comment (bsunanda) [reply Geometry/HGCalCommonData/plugins/dd4hep/DDHGCalTBModuleX.cc]: Same as last comment

**Comment by bsunanda** [comment]
@cmsbuild Please test

**Comment by cvuosalo** [comment]
@ianna: Are you satisfied with the changes in this PR?

**Comment by bsunanda** [comment]
@cmsbuild Please test

**Comment by ianna** [comment]
+1

**Comment by kpedro88** [comment]
+upgrade

**Comment by fabiocos** [comment]
+1

**Pull request merged**